keps/sig-auth/4872-harden-kubelet-cert-validation/README.md

```diff
@@ -0,0 +1,386 @@
+
+### Notes/Constraints/Caveats (Optional)
+
+When the kubelet requests a certificate through a CSR, it sets the CN to `system:node:<nodename>`, enforced by the admission controller as per [PR \#126015](https://github.com/kubernetes/kubernetes/pull/126015).
+
+However, certificates issued manually or through other mechanisms may not follow this convention.
+With the new validation, any certificate not following this `system:node:<nodename>` convention will be deemed invalid by the Kube API server.
+This will require cluster administrators to reissue any non-conforming certificates before enabling this feature.
+
+### Risks and Mitigations
+
+This could disrupt existing clusters that are using custom kubelet serving certificates.
+These clusters will need to reissue their certificates before enabling this feature. We will allow to disable the validation through a command-line flag to allow for a smooth transition.
+
+## Design Details
+
+### Enabling the feature
+
+We will introduce a feature flag `KubeletCertCNValidation` that will gate the usage of the new validation.
+This gate will start off by default in Alpha, will be turned on by default in Beta and will be removed in GA.
+
+In addition, we will allow to disable the validation through a command-line flag `--disable-kubelet-cert-cn-validation`.
+This flag can only be set if the `KubeletCertCNValidation` feature flag is enabled.
+This flag will allow cluster administrators to opt-out of this validation if they are using custom kubelet serving certificates that don't follow the `system:node:<nodename>` convention even after the feature gate is removed.
+
+#### Metrics
+
+In order to help cluster administrators determine if it's safe to enable the feature, we propose to add a new metric `kube_apiserver_validation_kubelet_cert_cn_errors` that will track the number of errors due to the new CN validation.
+If the feature gate is disabled, we will still add the validation code to the HTTP transport, however, if the validation fails we won't return an error, we will just increment the metric counter.
+In addition, we will log the error including the node name, so cluster administrators can identify which nodes are affected and need to reissue their certificates.
+
+We purposefully don't add the node name to the metric to avoid a high cardinality.
+The purpose of the metric is to easily/cheaply tell administrators if they can flip the feature on or not. If the answer is no (counter is greater than 0), the rest of the necessary information to detect the offending nodes will come from logs.
+
+Given that running the validation to feed the metric still has a cost, we won't run it if the validation is explicitly disabled with `--disable-kubelet-cert-cn-validation`.
+
+We will remove the metric once the feature is GA.
+
+> TODO: let's discuss this in the review. We could consider adding the node name to the metric or even keeping the metric post GA if it's valuable.
```

Comment:
node name has the cardinality problem, but leaving the metric sounds like a good thing to me, it can help to detect issues with the node certificates due to bugs per example